kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/FilesToColumnarBatchActionsIterator.java

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright (2023) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.kernel.internal.replay;
+
+import io.delta.kernel.client.FileReadContext;
+import io.delta.kernel.client.JsonHandler;
+import io.delta.kernel.client.ParquetHandler;
+import io.delta.kernel.client.TableClient;
+import io.delta.kernel.data.FileDataReadResult;
+import io.delta.kernel.expressions.Literal;
+import io.delta.kernel.fs.FileStatus;
+import io.delta.kernel.internal.util.InternalUtils;
+import io.delta.kernel.types.StructType;
+import io.delta.kernel.utils.CloseableIterator;
+import io.delta.kernel.utils.Tuple2;
+import io.delta.kernel.utils.Utils;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.*;
+
+/**
+ * This class takes as input a list of delta files (.json, .checkpoint.parquet) and produces an
+ * iterator of (FileDataReadResult, isFromCheckpoint) tuples.
+ *
+ * Users can pass in a `readSchema` to select which actions they want to consume.
+ */
+class FilesToColumnarBatchActionsIterator
+    implements CloseableIterator<Tuple2<FileDataReadResult, Boolean>> {
+
+    private final TableClient tableClient;
+
+    /**
+     * Iterator over the files.
+     *
+     * Each file will be split (by 1, or more) to yield an iterator of FileDataReadResults.
+     */
+    private final Iterator<FileStatus> filesIter;
+
+    private final StructType readSchema;
+
+    /**
+     * The current (FileDataReadResult, isFromCheckpoint) tuple. Whenever this iterator
+     * is exhausted, we will try and fetch the next one from the `filesIter`.
+     *
+     * If it is ever empty, that means there are no more batches to produce.
+     */
+    private Optional<CloseableIterator<Tuple2<FileDataReadResult, Boolean>>>
+        actionsIter;
+
+    private boolean closed;
+
+    public FilesToColumnarBatchActionsIterator(
+            TableClient tableClient,
+            List<FileStatus> files,
```

Comment:
nit: why not pass the LogSegment object here directly.

Reply:
Hm. Isn't it better to pass just what the class needs, and not more?

I would imagine that to test this, passing in just a list of files would be easier than having to "mock" (instantiate, create) a full LogSegment